tensornetwork/backends/jax/jitted_functions.py

```diff
@@ -58,84 +62,86 @@ def jax_lanczos(matvec: Callable, arguments: List, init: jax.ShapedArray,
         This should be used if `neig>1`.
+      precision: jax.lax.Precision type used in jax.numpy.vdot
     Returns:
       jax.numpy.ndarray: Eigenvalues
       list: Eigenvectors
     """
-
+    shape = init.shape
+    dtype = init.dtype
+    # TODO (mganahl): replace with restarted classical gram-schmidt
     def body_modified_gram_schmidt(i, vals):
       vector, krylov_vectors = vals
       v = krylov_vectors[i, :]
-      vector -= jax.numpy.vdot(
-          v, vector, precision=precision) * jax.numpy.reshape(v, vector.shape)
+      vector = vector - jax.numpy.vdot(v, vector, precision=precision) * v
       return [vector, krylov_vectors]
 
     def body_lanczos(vals):
-      current_vector, krylov_vectors, vector_norms = vals[0:3]
-      diagonal_elements, matvec, args, _ = vals[3:7]
-      threshold, i, maxiteration = vals[7:]
-      norm = jax.numpy.linalg.norm(current_vector)
-      normalized_vector = current_vector / norm
-      normalized_vector, krylov_vectors = jax.lax.cond(
-          reortho, True,
-          lambda x: jax.lax.fori_loop(0, i, body_modified_gram_schmidt,
-                                      [normalized_vector, krylov_vectors]),
-          False, lambda x: [normalized_vector, krylov_vectors])
-      Av = matvec(normalized_vector, *args)
-
-      diag_element = jax.numpy.vdot(normalized_vector, Av, precision=precision)
-
-      res = jax.numpy.reshape(
-          jax.numpy.ravel(Av) -
-          jax.numpy.ravel(normalized_vector) * diag_element -
-          krylov_vectors[i - 1] * norm, Av.shape)
-      krylov_vectors = jax.ops.index_update(krylov_vectors, jax.ops.index[i, :],
-                                            jax.numpy.ravel(normalized_vector))
-
-      vector_norms = jax.ops.index_update(vector_norms, jax.ops.index[i - 1],
-                                          norm)
-      diagonal_elements = jax.ops.index_update(diagonal_elements,
-                                               jax.ops.index[i - 1],
-                                               diag_element)
-
-      return [
-          res, krylov_vectors, vector_norms, diagonal_elements, matvec, args,
-          norm, threshold, i + 1, maxiteration
-      ]
+      krylov_vectors, alphas, betas, i = vals
+      previous_vector = krylov_vectors[i, :]
+
+      previous_vector, krylov_vectors = jax.lax.cond(
+          reortho,
+          lambda x: jax.lax.fori_loop(1, i, body_modified_gram_schmidt, x),
+          lambda x: x, [jax.numpy.ravel(previous_vector), krylov_vectors])
+      beta = jax.numpy.linalg.norm(previous_vector)
+      normalized_vector = previous_vector / beta
+      Av = matvec(jax.lax.reshape(normalized_vector, shape), *arguments)
+      alpha = jax.numpy.vdot(normalized_vector, Av, precision=precision)
+      alphas = alphas.at[i - 1].set(alpha)
+      betas = betas.at[i].set(beta)
+      next_vector = jax.lax.cond(
+          reortho, lambda x: Av, lambda x: jax.numpy.reshape(
```

Comment:
I think this would be easier to read if the False branch were defined separately above with a def

Reply:
done